vault/hcp_link/capabilities/meta/meta.go

```diff
@@ -189,3 +191,110 @@ func (h *hcpLinkMetaHandler) ListAuths(ctx context.Context, req *meta.ListAuthsR
 		Auths: auths,
 	}, nil
 }
+
+func (h *hcpLinkMetaHandler) GetClusterStatus(ctx context.Context, req *meta.GetClusterStatusRequest) (*meta.GetClusterStatusResponse, error) {
+	if h.wrappedCore.HAState() != consts.Active {
+		return nil, fmt.Errorf("node not active")
+	}
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+
+	haEnabled := h.wrappedCore.HAEnabled()
+	haStatus := &meta.HAStatus{
+		Enabled: haEnabled,
+	}
+
+	if haEnabled {
+		leader := &meta.HANode{
+			Hostname: hostname,
+		}
+
+		peers := h.wrappedCore.GetHAPeerNodesCached()
+
+		haNodes := make([]*meta.HANode, len(peers)+1)
+		haNodes[0] = leader
+
+		for i, peerNode := range peers {
+			haNodes[i+1] = &meta.HANode{
+				Hostname: peerNode.Hostname,
+			}
+		}
+
+		haStatus.Nodes = haNodes
+	}
+
+	raftStatus := &meta.RaftStatus{}
+	raftConfig, err := h.wrappedCore.GetRaftConfiguration(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if raftStatus != nil {
+		raftServers := make([]*meta.RaftServer, len(raftConfig.Servers))
+
+		var voterCount uint32
+		for i, srv := range raftConfig.Servers {
+			raftServers[i] = &meta.RaftServer{
+				NodeID:          srv.NodeID,
+				Address:         srv.Address,
+				Voter:           srv.Voter,
+				Leader:          srv.Leader,
+				ProtocolVersion: srv.ProtocolVersion,
+			}
+
+			if srv.Voter {
+				voterCount++
+			}
+		}
+
+		raftStatus.RaftConfiguration = &meta.RaftConfiguration{
+			Servers: raftServers,
+		}
+
+		quorumWarnings := make([]string, 0)
+		if voterCount == 0 {
+			quorumWarnings = append(quorumWarnings, "Only one server node found. Vault is not running in high availability mode.")
+		} else if voterCount%2 == 0 {
+			quorumWarnings = append(quorumWarnings, "Vault should have access to an odd number of voter nodes.")
+		}
+
+		if voterCount > 7 {
+			quorumWarnings = append(quorumWarnings, "Warning: Very large cluster detected.")
+		}
+
+		raftStatus.QuorumWarnings = quorumWarnings
+	}
+
+	raftAutopilotState, err := h.wrappedCore.GetRaftAutopilotState(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if raftAutopilotState != nil {
+		autopilotStatus := &meta.AutopilotStatus{
+			Healthy: raftAutopilotState.Healthy,
+		}
+
+		autopilotServers := make([]*meta.AutopilotServer, 0)
+		for _, srv := range raftAutopilotState.Servers {
+			autopilotServers = append(autopilotServers, &meta.AutopilotServer{
+				ID:      srv.ID,
+				Healthy: srv.Healthy,
+			})
+		}
+
+		raftStatus.AutopilotStatus = autopilotStatus
+	}
+
+	resp := &meta.GetClusterStatusResponse{
+		ClusterID:   h.wrappedCore.ClusterID(),
```

Comment:
I made a comment above about possible race conditions. Would this need to be protected against race conditions?

Reply:
How is this `GetClusterStatus` function going to be used? I am asking because HCP connection is established very early on even if Vault is sealed I think. So, if this function is called right away after connection, some info such as clusterID may not be even generated in Vault. It is worth testing this scenario. Though I think it is up to the caller to make sure it has all the info it needs, and I think the caller perhaps runs this on a cadence. So, this observation is really not a big deal. 
I just figured to raise this observation in case there is anything special to take into account.

Reply:
> How is this GetClusterStatus function going to be used? I am asking because HCP connection is established very early on even if Vault is sealed I think.

It will be consumed together with the update of node statuses -- every time the node statuses callback service is called by the link service to let the vault service know about the new statuses, the vault service will call the `GetClusterStatus` endpoint.